Fix processor cleanup on DagFileProcessorManager

pcolladosoto opened the pull request:
This PR fixes the cleanup procedures of the `DAGFileProcessor`s being spawned by the `DagFileProcessorManager` instance. Thus, this PR closes #22191.

We came across this bug when encountering a timeout triggering the forceful killing of these `DAGFileProcessor`s.

We have had somewhat lengthy discussion on an issue (#22191) which we encourage anyone to read for some more insight into the cause, discovery and posterior fix. People on that thread were extremely helpful! 😸

We **have not** included tests because we're unsure of how to test a behaviour such as this one. If pointed in the right direction we would be more than happy to add them. We can say however that we applied these changes on our own production Airflow instance and we haven't encountered the issue ever since.

On top of that we would be more than welcome if you made any suggestions to the code: for instance, when cleaning up a dictionary we're iterating over we decided to take note of the problematic `DAGFileProcessor`s to then remove them on a second `for` loop. Our background in programming is much stronger on some other languages and so we feel really uncomfortable pushing Python 'to the limit' in terms of relying on its implementation to make design choices. If there's anything that can be done better by all means say so.

Another fertile topic for discussion is how to `wait()` for the processes being killed through the `SIGKILL` signal. This has been brought up by @dlesco on #22191 and we agree with him on adding an optional timeout to the operation to avoid blocking in very bizarre circumstances (which the current solution would do). However, we decided to contribute our initial approach and then iterate on solutions within the PR.

Thanks a lot for letting me contribute to a tool with the expertise and size of Airflow: it's truly an honour.

Hope to make the merge as seamless as possible 😜

closes: #22191

pcolladosoto committed:
Fix processor cleanup

References to processors weren't being cleaned up after
killing them in the event of a timeout. This lead to
a crash caused by an unhandled exception when trying to
read from a closed end of a pipe.

pcolladosoto committed:
Fix processor cleanup

References to processors weren't being cleaned up after
killing them in the event of a timeout. This lead to
a crash caused by an unhandled exception when trying to
read from a closed end of a pipe.

pcolladosoto committed:
Reap the zombie when killing the processor

When calling `_kill_process()` we're generating
zombies which weren't being `wait()`ed for. This
led to a process leak we fix by just calling
`waitpid()` on the appropriate PIDs.

pcolladosoto committed:
Reap the zombie when killing the processor

When calling `_kill_process()` we're generating
zombies which weren't being `wait()`ed for. This
led to a process leak we fix by just calling
`waitpid()` on the appropriate PIDs.

potiuk commented on airflow/dag_processing/processor.py:
```diff
@@ -231,6 +231,9 @@ def _kill_process(self) -> None:
             os.kill(self._process.pid, signal.SIGKILL)
+
+            # Reap the spawned zombie
+            os.waitpid(self._process.pid, 0)
```
Indeed there might be a short time between even SIGKILL gets processed so waiting here makes sense.

malthe replied:
I don't know if there's anything here that could be racy, but there are some details here: https://bugs.python.org/issue42558.

potiuk replied:
Ough.  Very interesting . So it seems that this one (or similar) is safer (ad waitpid might crash for Python 3.9+) :) 

Do I read it right ? 

```
while self._process.poll() is None:
    sleep(0.001)
```

malthe replied:
Yes that is how I read it too.

@pcolladosoto you wanna make that change? I don't think I had anything else that needed changing.

pcolladosoto replied:
Hi @malthe! I just pushed the changes: I decided to import the `time` module so that I could call `time.sleep()` within the `while` loop. Hope I made the right choice... You can check out the changes on 27502be2e2a8db81. Thanks a lot for the input! This is something that I'm sure will come in handy for future projects 😸

potiuk commented on airflow/dag_processing/manager.py:
```diff
@@ -1073,6 +1074,14 @@ def _kill_timed_out_processors(self):
                 Stats.incr('dag_file_processor_timeouts')
                 processor.kill()
 
+                # Clean up processor references
```
This is really nice! Good catch.

pcolladosoto replied:
Thanks a ton! 😸

malthe commented on airflow/dag_processing/manager.py:
```diff
@@ -1073,6 +1074,14 @@ def _kill_timed_out_processors(self):
+
+        # Clean up `self._processors` after iterating over it
+        for proc in processors_to_remove:
+            self._processors.pop(proc)
```
Or `del self._processors[proc]`.

potiuk replied:
Why ?

malthe replied:
I guess it communicates better, but I noticed afterwards that the rest of the code also uses `pop` everywhere.

pcolladosoto committed:
Reap resulting zombies in a safe way

According to @potiuk's and @malthe's input, the way
we were reaping the zombies could cause some racy and
unwanted situations. As seen on the discussion over at
`https://bugs.python.org/issue42558` we can safely
reap the spawned zombies with the changes we have
introduced.

pcolladosoto committed:
Reap resulting zombies in a safe way

According to @potiuk's and @malthe's input, the way
we were reaping the zombies could cause some racy and
unwanted situations. As seen on the discussion over at
`https://bugs.python.org/issue42558` we can safely
reap the spawned zombies with the changes we have
introduced.

potiuk commented on airflow/dag_processing/processor.py:
```diff
@@ -231,6 +232,10 @@ def _kill_process(self) -> None:
             self.log.warning("Killing DAGFileProcessorProcess (PID=%d)", self._process.pid)
             os.kill(self._process.pid, signal.SIGKILL)
+
+            # Reap the spawned zombie
```
```suggestion
            # Reap the spawned zombie. We active wait, becasue in Python 3.9 `waitpid` might lead to an
            # exception, due to change in Python standard library and possibility of race condition
            # see https://bugs.python.org/issue42558 
```

potiuk replied:
Small suggestion - to capture the discussion above. This is basically the type of comment for the "future me". I (and we all) will likely forget why this is done this way and we might atempt to fix the active waiting :)

potiuk replied:
BTW - one more thought. I actually had - not too often, but at least few times in my life (hardware problem) - where sigkill did not actually work and process was running indefinitely (even though sig kill is esentially the "should always work" case). But maybe some "what-if" discussion might be useful (XKCD reference accidental).

Should we handle this case? Also what happnes if (this is unlikely because we started the process) if we have no permission to send sigkill to the process? Not sure about scenario that can lead to it (changing effecitve group of the user? . Do you think we should handle it?

This is a bit of "academic" discussion, but if we are going so deeply that we are active-waiting for the sigkilled process, I thought we might at least spend a bit of time on pondering on various super-edge cases :)

pcolladosoto committed:
Explain why we are actively waiting

As suggested by @potiuk explaining why we chose to actively wait on an scenario such as this one can indeed be useful for anybody taking a look at the code some time from now...

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

pcolladosoto committed:
Explain why we are actively waiting

As suggested by @potiuk explaining why we chose to actively wait on an scenario such as this one can indeed be useful for anybody taking a look at the code some time from now...

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

pcolladosoto committed:
Fix small typo and triling whitespace

After accepting the changes proposed on the PR
we found a small typo (we make those on a daily basis)
and a trailing whitespace we though was nice to delete.
Hope we made the right choice!

pcolladosoto committed:
Fix small typo and triling whitespace

After accepting the changes proposed on the PR
we found a small typo (we make those on a daily basis)
and a trailing whitespace we though was nice to delete.
Hope we made the right choice!

pcolladosoto committed:
Fix call to `poll()`

We were calling `poll()` through the `_process` attribute
and, as shown on the static checks triggered by GitHub,
it's not defined for the `BaseProcess` class. We instead
have to call `poll()` through `BaseProcess`'s `_popen`
attribute.

pcolladosoto committed:
Fix call to `poll()`

We were calling `poll()` through the `_process` attribute
and, as shown on the static checks triggered by GitHub,
it's not defined for the `BaseProcess` class. We instead
have to call `poll()` through `BaseProcess`'s `_popen`
attribute.

pcolladosoto committed:
Merge branch 'main' of origin

Some changes have been force pushed!

pcolladosoto committed:
Prevent static check from failing

After reading through `multiprocessing`'s implementation we
really didn't know why the static check on line `239` was
failing: the process should contain a `_popen` attribute...
That's when we found line `223` and discovered the trailing
`# type: ignore` comment. After reading up on it we found
that it instructs *MyPy* not to statically check that very
line. Given we're having trouble with the exact same attribute
we decided to include the same directive for the static checker.
Hope we made the right call!

pcolladosoto committed:
Fix test for `_kill_timed_out_processors()`

We hadn't updated the tests for the method whose
body we've altered. This caused the tests to fail
when trying to retrieve a processor's *waitable*,
a property similar to a *file descriptor* in
UNIX-like systems. We have added a mock property to
the `processor` and we've also updated the `manager`'s
attributes so as to faithfully recreate the state of
the data sctructures at a moment when a `processor`
is to be terminated.

Please note the `assertions` at the end are meant to
check we reach the `manager`'s expected state. We have
chosen to check the number of processor's against an
explicit value because we're defining `manager._processors`
explicitly within the test. On the other hand, `manager.waitables`
can have a different length depending on the call to
`DagFileProcessorManager`'s `__init__()`. In this test the
expected initial length is `1` given we're passing `MagicMock()`
as the `signal_conn` when instantiating the manager. However,
if this were to be changed the tests would 'inexplicably' fail.
Instead of checking `manager.waitables`' length against a hardcoded
value we decided to instead compare it to its initial length
so as to emphasize we're interested in the change in length, not
its absolute value.

pcolladosoto committed:
Fix `black` checks and `mock` decorators

One of the methods we are to mock required a rather
long `@mock.patch` decorator which didn't pass the
checks made by `black` on the precommit hooks. On
top of that, we messed up the ordering of the
`@mock.patch` decorators which meant we didn't
set them up properly. This manifested as a `KeyError`
on the method we're currently testing. O_o
